File: src/promaster_dash/services/trip_manager.py
"""Automatic trip detection and statistics tracking."""

from __future__ import annotations

import time
from enum import Enum
from typing import Optional

from PySide6.QtCore import QObject, Signal, Slot

from promaster_dash.models.data_records import GPSData, TripStats
from promaster_dash.models.vehicle_state import VehicleState, TRANS_WARN, COOL_WARN
from promaster_dash.services.logging_service import LoggingService


class TripState(Enum):
    """Trip state machine states."""

    IDLE = "idle"  # No active trip
    ACTIVE = "active"  # Trip in progress
    ENDING = "ending"  # Vehicle stopped, waiting to confirm trip end


class TripManager(QObject):
    """
    Manages automatic trip detection and statistics.

    State machine:
        IDLE ──(speed > 5 mph)──> ACTIVE
        ACTIVE ──(speed < 1 mph)──> ENDING
        ENDING ──(speed > 5 mph)──> ACTIVE
        ENDING ──(5 min elapsed)──> IDLE (emit trip_ended)
    """

    # Signals
    trip_started = Signal(int)  # trip_id
    trip_ended = Signal(object)  # TripStats
    trip_stats_updated = Signal(object)  # TripStats (periodic updates)
    state_changed = Signal(str)  # TripState value

    # Configuration
    START_SPEED_THRESHOLD = 5.0  # mph to start trip
    STOP_SPEED_THRESHOLD = 1.0  # mph considered stopped
    STOP_TIMEOUT_SECS = 300  # 5 minutes stationary to end trip
    STATS_UPDATE_INTERVAL = 5.0  # seconds between stats updates

    def __init__(
        self,
        logging_service: Optional[LoggingService] = None,
    ):
        super().__init__()
        self._logging_service = logging_service
        self._state = TripState.IDLE
        self._stats: Optional[TripStats] = None

        # Timing
        self._stop_start_ts: Optional[float] = None  # When vehicle first stopped
        self._last_update_ts = 0.0
        self._last_stats_emit_ts = 0.0

    @property
    def state(self) -> TripState:
        """Current trip state."""
        return self._state

    @property
    def stats(self) -> Optional[TripStats]:
        """Current trip statistics (None if no active trip)."""
        return self._stats

    @property
    def is_active(self) -> bool:
        """True if a trip is currently active or ending."""
        return self._state in (TripState.ACTIVE, TripState.ENDING)

    @Slot(object)
    def on_gps_update(self, data: GPSData) -> None:
        """
        Process a GPS update.

        Args:
            data: GPS data from GPSService
        """
        if not data.fix_ok:
            return

        now = time.time()
        dt = now - self._last_update_ts if self._last_update_ts > 0 else 0
        self._last_update_ts = now

        speed = data.speed_mph or 0

        if self._state == TripState.IDLE:
            self._handle_idle(speed, now)
        elif self._state == TripState.ACTIVE:
            self._handle_active(data, speed, now, dt)
        elif self._state == TripState.ENDING:
            self._handle_ending(data, speed, now, dt)

    def on_vehicle_state_update(self, vehicle_state: VehicleState, dt: float) -> None:
        """
        Update trip stats with vehicle telemetry.

        Call this periodically (e.g., every UI tick) to track temperatures.

        Args:
            vehicle_state: Current vehicle state
            dt: Time delta since last call
        """
        if self._stats is None:
            return

        self._stats.update_temps(
            vehicle_state.trans_f,
            vehicle_state.coolant_f,
            dt,
            TRANS_WARN,
            COOL_WARN,
        )

    def _handle_idle(self, speed: float, now: float) -> None:
        """Handle state transitions from IDLE."""
        if speed >= self.START_SPEED_THRESHOLD:
            self._start_trip(now)

    def _handle_active(self, data: GPSData, speed: float, now: float, dt: float) -> None:
        """Handle state transitions and updates in ACTIVE state."""
        # Update trip stats
        if data.lat is not None and data.lon is not None:
            self._stats.update_distance(data.lat, data.lon)

        if speed >= 0:
            self._stats.update_speed(speed)

        # Check for stop
        if speed < self.STOP_SPEED_THRESHOLD:
            self._transition_to(TripState.ENDING)
            self._stop_start_ts = now
        else:
            # Still moving, emit periodic stats
            self._maybe_emit_stats(now)

    def _handle_ending(self, data: GPSData, speed: float, now: float, dt: float) -> None:
        """Handle state transitions in ENDING state."""
        # Still update stats while ending
        if data.lat is not None and data.lon is not None:
            self._stats.update_distance(data.lat, data.lon)

        if speed >= self.START_SPEED_THRESHOLD:
            # Resumed moving
            self._transition_to(TripState.ACTIVE)
            self._stop_start_ts = None
        elif self._stop_start_ts and (now - self._stop_start_ts) >= self.STOP_TIMEOUT_SECS:
            # Been stopped long enough, end trip
            self._end_trip(now)
        else:
            # Still waiting
            self._maybe_emit_stats(now)

    def _start_trip(self, now: float) -> None:
        """Start a new trip."""
        self._stats = TripStats(start_ts=now)

        # Create database record
        if self._logging_service:
            trip_id = self._logging_service.create_trip(now)
            if trip_id:
                self._stats.trip_id = trip_id

        self._transition_to(TripState.ACTIVE)
        self._stop_start_ts = None

        if self._stats.trip_id:
            self.trip_started.emit(self._stats.trip_id)

    def _end_trip(self, now: float) -> None:
        """End the current trip."""
        if self._stats is None:
            self._transition_to(TripState.IDLE)
            return

        self._stats.end_ts = now

        # Finalize database record
        if self._logging_service and self._stats.trip_id:
            self._logging_service.finalize_trip(self._stats.trip_id, self._stats)

        # Emit final stats
        self.trip_ended.emit(self._stats)

        # Reset
        self._stats = None
        self._stop_start_ts = None
        self._transition_to(TripState.IDLE)

    def _transition_to(self, new_state: TripState) -> None:
        """Transition to a new state."""
        if new_state != self._state:
            self._state = new_state
            self.state_changed.emit(new_state.value)

    def _maybe_emit_stats(self, now: float) -> None:
        """Emit stats update if interval has passed."""
        if self._stats and (now - self._last_stats_emit_ts) >= self.STATS_UPDATE_INTERVAL:
            self.trip_stats_updated.emit(self._stats)
            self._last_stats_emit_ts = now

    def force_end_trip(self) -> None:
        """Force-end the current trip (e.g., on app shutdown)."""
        if self._state != TripState.IDLE and self._stats:
            self._end_trip(time.time())

    def reset(self) -> None:
        """Reset to idle state without logging."""
        self._stats = None
        self._stop_start_ts = None
        self._state = TripState.IDLE
        self.state_changed.emit(TripState.IDLE.value)

File: mobile/src/services/tripManager.ts
/**
 * Automatic trip detection and statistics tracking.
 * Direct port of Python TripManager state machine.
 *
 * State machine:
 *   IDLE ──(speed > 5 mph)──> ACTIVE
 *   ACTIVE ──(speed < 1 mph)──> ENDING
 *   ENDING ──(speed > 5 mph)──> ACTIVE
 *   ENDING ──(5 min elapsed)──> IDLE (trip ends)
 */

import AsyncStorage from '@react-native-async-storage/async-storage';
import type { GPSData } from '../models/types';
import { TRANS_WARN, COOL_WARN } from '../models/types';
import { TripStats } from '../models/TripStats';
import {
  createTrip,
  finalizeTrip,
  logBreadcrumb,
} from './loggingService';
import { useVehicleStore } from '../store/vehicleStore';
import { dlog } from './debugLog';

export type TripState = 'idle' | 'active' | 'ending';

// Configuration
let START_SPEED_THRESHOLD = 5.0;
let STOP_SPEED_THRESHOLD = 1.0;
let STOP_TIMEOUT_SECS = 600; // 10 minutes (was 5 — increased to prevent fragmentation at traffic stops)

// How many consecutive slow readings required before entering ENDING state
const SLOW_COUNT_REQUIRED = 3;

// State
let state: TripState = 'idle';
let stats: TripStats | null = null;
let stopStartTs: number | null = null;
let lastUpdateTs = 0;
let lastStatsEmitTs = 0;
let slowCount = 0; // consecutive below-threshold readings
let isRestoring = false;

const STATS_UPDATE_INTERVAL = 5.0;

// --- Trip state persistence ---

const TRIP_STATE_KEY = '@promaster/activeTripState';

interface PersistedTripState {
  tripId: number;
  startTs: number;
  distanceMi: number;
  maxTransF: number | null;
  maxCoolantF: number | null;
  transWarnSecs: number;
  coolantWarnSecs: number;
  speedSamples: number;
  speedSum: number;
  lastLat: number | null;
  lastLon: number | null;
}

/** Save active trip state to AsyncStorage. Called periodically. */
export async function persistTripState(): Promise<void> {
  if (!stats || !stats.tripId) return;
  const data: PersistedTripState = {
    tripId: stats.tripId,
    startTs: stats.startTs,
    distanceMi: stats.distanceMi,
    maxTransF: stats.maxTransF,
    maxCoolantF: stats.maxCoolantF,
    transWarnSecs: stats.transWarnSecs,
    coolantWarnSecs: stats.coolantWarnSecs,
    speedSamples: stats.speedSamples,
    speedSum: stats.speedSum,
    lastLat: stats.lastLat,
    lastLon: stats.lastLon,
  };
  try {
    await AsyncStorage.setItem(TRIP_STATE_KEY, JSON.stringify(data));
  } catch (e: any) {
    dlog(`Trip: Failed to persist state: ${e.message}`);
  }
}

/** Clear persisted trip state (on normal trip end). */
async function clearPersistedTripState(): Promise<void> {
  try {
    await AsyncStorage.removeItem(TRIP_STATE_KEY);
  } catch (e: any) {
    dlog(`Trip: Failed to clear persisted state: ${e.message}`);
  }
}

/**
 * Try to restore an active trip from AsyncStorage.
 * Call on app startup before GPS begins.
 * Returns true if a trip was restored.
 */
export async function restoreTrip(): Promise<boolean> {
  isRestoring = true;
  try {
    const raw = await AsyncStorage.getItem(TRIP_STATE_KEY);
    if (!raw) return false;

    const data: PersistedTripState = JSON.parse(raw);

    // Discard trips older than 24 hours — clearly not an ongoing drive.
    const ageHours = (Date.now() / 1000 - data.startTs) / 3600;
    if (ageHours > 24) {
      dlog(`Trip: Persisted trip is ${ageHours.toFixed(0)}h old, discarding`);
      await clearPersistedTripState();
      return false;
    }

    // Restore TripStats
    stats = new TripStats();
    stats.tripId = data.tripId;
    stats.startTs = data.startTs;
    stats.distanceMi = data.distanceMi;
    stats.maxTransF = data.maxTransF;
    stats.maxCoolantF = data.maxCoolantF;
    stats.transWarnSecs = data.transWarnSecs;
    stats.coolantWarnSecs = data.coolantWarnSecs;
    stats.speedSamples = data.speedSamples;
    stats.speedSum = data.speedSum;
    stats.lastLat = data.lastLat;
    stats.lastLon = data.lastLon;

    // Resume in active state
    state = 'active';
    stopStartTs = null;
    slowCount = 0;

    // Update store
    const store = useVehicleStore.getState();
    store.setTripState(true, stats.tripId, stats.startTs);
    store.updateTripDistance(stats.distanceMi);

    dlog(`Trip: Restored active trip #${stats.tripId} (started ${new Date(stats.startTs * 1000).toLocaleTimeString()}, ${stats.distanceMi.toFixed(1)} mi)`);
    return true;
  } catch (e: any) {
    dlog(`Trip: Failed to restore: ${e.message}`);
    await clearPersistedTripState();
    return false;
  } finally {
    isRestoring = false;
  }
}

/** Get current trip state. */
export function getTripState(): TripState {
  return state;
}

/** Get current trip stats. */
export function getTripStats(): TripStats | null {
  return stats;
}

/** Update configuration thresholds. */
export function setTripConfig(config: {
  startSpeed?: number;
  stopTimeout?: number;
}): void {
  if (config.startSpeed !== undefined) START_SPEED_THRESHOLD = config.startSpeed;
  if (config.stopTimeout !== undefined) STOP_TIMEOUT_SECS = config.stopTimeout;
}

/** Process a GPS update — drives the state machine. */
export async function onGpsUpdate(data: GPSData): Promise<void> {
  if (isRestoring) return;
  if (!data.fixOk) return;

  const now = Date.now() / 1000;
  const dt = lastUpdateTs > 0 ? now - lastUpdateTs : 0;
  lastUpdateTs = now;

  const speed = data.speedMph ?? 0;

  if (state === 'idle') {
    await handleIdle(speed, now);
  } else if (state === 'active') {
    await handleActive(data, speed, now);
  } else if (state === 'ending') {
    await handleEnding(data, speed, now);
  }
}

/** Update trip stats with OBD temperature data. Call periodically. */
export function onVehicleStateUpdate(
  transF: number | null,
  coolantF: number | null,
  dt: number,
): void {
  if (!stats) return;
  stats.updateTemps(transF, coolantF, dt, TRANS_WARN, COOL_WARN);
}

async function handleIdle(speed: number, now: number): Promise<void> {
  if (speed >= START_SPEED_THRESHOLD) {
    await startTrip(now);
  }
}

async function handleActive(
  data: GPSData,
  speed: number,
  now: number,
): Promise<void> {
  if (!stats) return;

  // Update distance
  if (data.lat !== null && data.lon !== null) {
    stats.updateDistance(data.lat, data.lon);
  }

  if (speed >= 0) {
    stats.updateSpeed(speed);
  }

  // Log breadcrumb
  await maybeBreadcrumb(data);

  // Require SLOW_COUNT_REQUIRED consecutive slow readings before entering ENDING
  // This prevents brief GPS glitches or traffic light stops from fragmenting trips
  if (speed < STOP_SPEED_THRESHOLD) {
    slowCount++;
    if (slowCount >= SLOW_COUNT_REQUIRED) {
      transitionTo('ending');
      stopStartTs = now;
      slowCount = 0;
    }
  } else {
    slowCount = 0;
    maybeEmitStats(now);
  }
}

async function handleEnding(
  data: GPSData,
  speed: number,
  now: number,
): Promise<void> {
  if (!stats) return;

  // Still update distance
  if (data.lat !== null && data.lon !== null) {
    stats.updateDistance(data.lat, data.lon);
  }

  // Log breadcrumb
  await maybeBreadcrumb(data);

  if (speed >= START_SPEED_THRESHOLD) {
    // Resumed driving
    transitionTo('active');
    stopStartTs = null;
    slowCount = 0;
  } else if (stopStartTs && now - stopStartTs >= STOP_TIMEOUT_SECS) {
    // Been stopped long enough — end trip
    await endTrip(now);
  } else {
    maybeEmitStats(now);
  }
}

async function startTrip(now: number): Promise<void> {
  stats = new TripStats();
  stats.startTs = now;

  const tripId = await createTrip(now);
  if (tripId) {
    stats.tripId = tripId;
  }

  transitionTo('active');
  stopStartTs = null;

  // Update store
  const store = useVehicleStore.getState();
  store.setTripState(true, stats.tripId, now);

  persistTripState(); // fire-and-forget
}

async function endTrip(now: number): Promise<void> {
  await clearPersistedTripState();

  if (!stats) {
    transitionTo('idle');
    return;
  }

  stats.endTs = now;

  if (stats.tripId) {
    await finalizeTrip(stats.tripId, stats);
  }

  // Update store
  const store = useVehicleStore.getState();
  store.updateTripDistance(stats.distanceMi);
  store.setTripState(false, null, null);

  stats = null;
  stopStartTs = null;
  transitionTo('idle');
}

/** Force-end trip (e.g., on app shutdown). */
export async function forceEndTrip(): Promise<void> {
  if (state !== 'idle' && stats) {
    await endTrip(Date.now() / 1000);
  }
}

/** Reset to idle without logging. */
export function reset(): void {
  stats = null;
  stopStartTs = null;
  state = 'idle';
  clearPersistedTripState();
}

function transitionTo(newState: TripState): void {
  state = newState;
}

function maybeEmitStats(now: number): void {
  if (stats && now - lastStatsEmitTs >= STATS_UPDATE_INTERVAL) {
    // Update store with latest trip distance
    const store = useVehicleStore.getState();
    store.updateTripDistance(stats.distanceMi);
    lastStatsEmitTs = now;
    persistTripState(); // fire-and-forget
  }
}

async function maybeBreadcrumb(data: GPSData): Promise<void> {
  if (!stats?.tripId || data.lat === null || data.lon === null) return;

  const store = useVehicleStore.getState();
  await logBreadcrumb({
    tripId: stats.tripId,
    ts: Date.now() / 1000,
    lat: data.lat,
    lon: data.lon,
    elevationFt: data.elevationFt,
    speedMph: data.speedMph,
    headingDeg: data.headingDeg,
    transF: store.transF,
    coolantF: store.coolantF,
    voltageV: store.voltageV,
    gradePct: data.gradePct,
    obdSpeedMph: store.obdSpeedMph,
  });
}
